File: app/src/main/java/com/yeraygarcia/recipes/adapter/EditRecipeAdapter.java
package com.yeraygarcia.recipes.adapter;

import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.widget.TextViewCompat;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableString;
import android.text.style.UnderlineSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.URLUtil;
import android.widget.ImageView;
import android.widget.TextView;

import com.yeraygarcia.recipes.EditItemDialogFragment;
import com.yeraygarcia.recipes.R;
import com.yeraygarcia.recipes.database.entity.RecipeStep;
import com.yeraygarcia.recipes.database.entity.custom.UiRecipe;
import com.yeraygarcia.recipes.database.entity.custom.UiRecipeIngredient;
import com.yeraygarcia.recipes.util.Debug;
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.Locale;
import java.util.UUID;

import static com.yeraygarcia.recipes.EditItemDialogFragment.TAG_FRAGMENT_EDIT_ITEM_DIALOG;

public class EditRecipeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEWTYPE_DURATION_SOURCE = 1;
    private static final int VIEWTYPE_INGREDIENTS_HEADER = 2;
    private static final int VIEWTYPE_STEPS_HEADER = 3;
    private static final int VIEWTYPE_INGREDIENT = 4;
    private static final int VIEWTYPE_STEP = 5;

    private FragmentActivity mParentActivity;
    private RecipeDetailViewModel mViewModel;
    private LayoutInflater mInflater;

    private UiRecipe mUiRecipe;
    private List<UiRecipeIngredient> mIngredients;
    private List<RecipeStep> mSteps;

    private int mSelectedStep = RecyclerView.NO_POSITION;

    private Integer mItemCount;

    // Constructors

    public EditRecipeAdapter(FragmentActivity parent, RecipeDetailViewModel viewModel) {
        Debug.d(this, "RecipeDetailAdapter(parent, viewModel)");

        mParentActivity = parent;
        mViewModel = viewModel;
        mInflater = LayoutInflater.from(parent);
    }

    // Internal classes

    class IngredientsHeaderViewHolder extends RecyclerView.ViewHolder {

        ImageView itemDecrease;
        ImageView itemIncrease;
        TextView itemServings;

        private IngredientsHeaderViewHolder(View itemView) {
            super(itemView);
            itemDecrease = itemView.findViewById(R.id.imageview_decrease_servings);
            itemIncrease = itemView.findViewById(R.id.imageview_increase_servings);
            itemServings = itemView.findViewById(R.id.textview_servings);

            itemDecrease.setOnClickListener(v -> {
                mUiRecipe.getRecipe().decreasePortions();
                mViewModel.update(mUiRecipe.getRecipe());
            });

            itemIncrease.setOnClickListener(v -> {
                mUiRecipe.getRecipe().increasePortions();
                mViewModel.update(mUiRecipe.getRecipe());
            });
        }

    }

    class HeaderViewHolder extends RecyclerView.ViewHolder {

        TextView itemTitle;

        private HeaderViewHolder(View itemView) {
            super(itemView);
            itemTitle = itemView.findViewById(R.id.textview_title);
        }

    }

    class DurationSourceViewHolder extends RecyclerView.ViewHolder {

        ImageView itemDurationIcon;
        ImageView itemSourceIcon;

        TextView itemDuration;
        TextView itemSource;

        private DurationSourceViewHolder(View itemView) {
            super(itemView);
            itemDurationIcon = itemView.findViewById(R.id.imageview_duration_icon);
            itemSourceIcon = itemView.findViewById(R.id.imageview_source_icon);
            itemDuration = itemView.findViewById(R.id.textview_duration);
            itemSource = itemView.findViewById(R.id.textview_source);
            itemSource.setOnClickListener(v -> {
                String urlString = (String) v.getTag();

                if (URLUtil.isValidUrl(urlString)) {
                    Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(urlString));
                    mParentActivity.startActivity(browserIntent);
                }
            });
        }

    }

    class IngredientsViewHolder extends RecyclerView.ViewHolder {

        TextView itemIngredientName;

        private IngredientsViewHolder(View itemView) {
            super(itemView);
            itemIngredientName = itemView.findViewById(R.id.textview_ingredient_name);

            itemIngredientName.setOnClickListener(view -> {
                UUID shoppingListItemId = mIngredients.get(getIngredientPosition(getAdapterPosition())).getId();
                showEditDialog(shoppingListItemId);
            });
        }
    }

    private void showEditDialog(UUID shoppingListItemId) {
        DialogFragment dialog = EditItemDialogFragment.newInstance(shoppingListItemId);
        dialog.show(mParentActivity.getSupportFragmentManager(), TAG_FRAGMENT_EDIT_ITEM_DIALOG);
    }

//    private void saveQuantityToDraft(String newQuantityText, int position) {
//        Debug.d(this, "saveQuantityToDraft(newQuantityText = " + newQuantityText + ", position = " + position + ")");
//        if (position == RecyclerView.NO_POSITION) {
//            return;
//        }
//
//        UiRecipeIngredient ingredient = mIngredients.get(getIngredientPosition(position));
//        Double oldUnitQuantity = ingredient.getQuantity();
//        double portions = ingredient.getPortions();
//        Double oldQuantity = oldUnitQuantity == null ? null : oldUnitQuantity * portions;
//
//        try {
//            if (newQuantityText.isEmpty() && oldQuantity != null
//                    || !newQuantityText.isEmpty() && oldQuantity == null
//                    || !Double.valueOf(newQuantityText).equals(oldQuantity)) {
//                // value has changed
//                Double newQuantity = newQuantityText.isEmpty() ? null : Double.valueOf(newQuantityText) / portions;
//                ingredient.setQuantity(newQuantity);
//                mViewModel.saveDraft(ingredient);
//            }
//        } catch (NumberFormatException ignored) {
//            Debug.d(this, "The entered value isn't a valid number");
//        }
//    }

    class StepsViewHolder extends RecyclerView.ViewHolder {

        TextView itemStepNumber;
        TextView itemStepDescription;

        private StepsViewHolder(View itemView) {
            super(itemView);
            itemStepNumber = itemView.findViewById(R.id.textview_step_number);
            itemStepDescription = itemView.findViewById(R.id.textview_step_description);

            itemView.setOnClickListener(v -> {
                final int position = getAdapterPosition();

                // updating old position
                notifyItemChanged(mSelectedStep);
                // deselect current item if it's already selected and the user clicks again on it
                mSelectedStep = (position == mSelectedStep) ? RecyclerView.NO_POSITION : position;
                // update new position
                notifyItemChanged(mSelectedStep);
            });
        }

    }

    // Overrides

    @Override
    public int getItemViewType(int position) {
        //Debug.d(this, "getItemViewType(position=" + String.valueOf(position) + ")");

        // there is always at least one ingredient item to show the "no ingredients" when empty
        final int ingredientsCount = (mIngredients == null) ? 1 : Math.max(mIngredients.size(), 1);

        if (position == 0) {
            return VIEWTYPE_DURATION_SOURCE;
        } else if (position == 1) {
            return VIEWTYPE_INGREDIENTS_HEADER;
        } else if (position < ingredientsCount + 2) {
            return VIEWTYPE_INGREDIENT;
        } else if (position == ingredientsCount + 2) {
            return VIEWTYPE_STEPS_HEADER;
        } else {
            return VIEWTYPE_STEP;
        }
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //Debug.d(this, "onCreateViewHolder(parent, viewType=" + String.valueOf(viewType) + ")");

        View view;
        switch (viewType) {
            case VIEWTYPE_INGREDIENTS_HEADER:
                view = mInflater.inflate(R.layout.item_header_ingredients, parent, false);
                return new IngredientsHeaderViewHolder(view);

            case VIEWTYPE_STEPS_HEADER:
                view = mInflater.inflate(R.layout.item_header_generic, parent, false);
                return new HeaderViewHolder(view);

            case VIEWTYPE_DURATION_SOURCE:
                view = mInflater.inflate(R.layout.item_duration_source, parent, false);
                return new DurationSourceViewHolder(view);

            case VIEWTYPE_INGREDIENT:
                view = mInflater.inflate(R.layout.item_edit_ingredient, parent, false);
                return new IngredientsViewHolder(view);

            default:
                view = mInflater.inflate(R.layout.item_step, parent, false);
                return new StepsViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        //Debug.d(this, "onBindViewHolder(holder, position=" + String.valueOf(position) + ")");

        switch (holder.getItemViewType()) {

            case VIEWTYPE_INGREDIENTS_HEADER:
                if (mUiRecipe != null) {
                    String portions = String.format(Locale.getDefault(), "%d", mUiRecipe.getRecipe().getPortions());
                    ((IngredientsHeaderViewHolder) holder).itemServings.setText(portions);

                    int decreaseVisibility = mUiRecipe.getRecipe().getPortions() > 1 ? View.VISIBLE : View.GONE;
                    ((IngredientsHeaderViewHolder) holder).itemDecrease.setVisibility(decreaseVisibility);
                }
                break;

            case VIEWTYPE_STEPS_HEADER:
                ((HeaderViewHolder) holder).itemTitle.setText(R.string.header_steps);
                break;

            case VIEWTYPE_DURATION_SOURCE:
                if (mUiRecipe != null) {
                    formatDurationSourceView((DurationSourceViewHolder) holder,
                            mUiRecipe.getRecipe().getDurationInMinutes(),
                            mUiRecipe.getRecipe().getUrl());
                }
                break;

            case VIEWTYPE_INGREDIENT:
                IngredientsViewHolder viewHolder = (IngredientsViewHolder) holder;
                if (mIngredients != null && mIngredients.size() > 0) {
                    UiRecipeIngredient ingredient = mIngredients.get(getIngredientPosition(position));
//                    UiRecipeIngredient ingredientDraft = mViewModel.getRecipeIngredientsDraft().get(ingredient.getId());
//                    if (ingredientDraft != null) {
//                        ingredient = ingredientDraft;
//                    }
                    viewHolder.itemIngredientName.setText(ingredient.toString());
                } else {
                    Debug.d(this, "no ingredients");
                    // Covers the case of data not being ready yet.
                    viewHolder.itemIngredientName.setText(R.string.no_ingredients);
                }
                break;

            default:
                holder.itemView.setSelected(mSelectedStep == position);
                if (mSteps != null && mSteps.size() > 0) {
                    final RecipeStep currentStep = mSteps.get(getStepPosition(position));
                    ((StepsViewHolder) holder).itemStepNumber.setText(formatSortOrder(currentStep.getSortOrder()));
                    ((StepsViewHolder) holder).itemStepDescription.setText(currentStep.getDescription());
                } else {
                    // Covers the case of data not being ready yet.
                    ((StepsViewHolder) holder).itemStepDescription.setText(R.string.no_steps);
                }
                break;
        }
    }

    @Override
    public int getItemCount() {
        //Debug.d(this, "getItemCount() -> " + size);
        if (mItemCount == null) {
            calculateItemCount();
        }
        return mItemCount;
    }

    private void calculateItemCount() {
        mItemCount = 0;
        if (mUiRecipe != null) {
            mItemCount += 3;
        }
        if (mIngredients != null) {
            // there is always at least one ingredient item to show the "no ingredients" when empty
            mItemCount += Math.max(mIngredients.size(), 1);
        }
        if (mSteps != null) {
            // there is always at least one step item to show the "no steps" when empty
            mItemCount += Math.max(mSteps.size(), 1);
        }
    }

    private int getIngredientPosition(int position) {
        return position - 2;
    }

    private int getStepPosition(int position) {
        int ingredientsSize = mIngredients == null ? 0 : mIngredients.size();
        return position - ingredientsSize - 3;
    }

    // Getters and Setters

    public void setRecipe(UiRecipe recipe) {
        Debug.d(this, "setRecipe(recipe)");
        mUiRecipe = recipe;
        mSteps = recipe.getSteps();
        calculateItemCount();
        notifyDataSetChanged();
    }

    public void setIngredients(List<UiRecipeIngredient> ingredients) {
        Debug.d(this, "setIngredients(ingredients(" + ingredients.size() + "))");
        mIngredients = ingredients;
        calculateItemCount();
        notifyDataSetChanged();
    }

    // Methods

    private String formatSortOrder(int sortOrder) {
        return String.format(Locale.getDefault(), "%d", sortOrder);
    }

    private String formatDuration(long duration) {
        return mParentActivity.getString(R.string.duration_format, String.format(Locale.getDefault(), "%d", duration));
    }

    private void formatDurationSourceView(DurationSourceViewHolder viewHolder,
                                          long duration, String url) {

        viewHolder.itemDuration.setText(formatDuration(duration));

        if (url == null) {
            viewHolder.itemSourceIcon.setVisibility(View.INVISIBLE);
            viewHolder.itemSource.setVisibility(View.INVISIBLE);
            viewHolder.itemSource.setTag(null);
        } else {
            viewHolder.itemSourceIcon.setVisibility(View.VISIBLE);
            viewHolder.itemSource.setVisibility(View.VISIBLE);
            viewHolder.itemSource.setTag(url);

            if (URLUtil.isValidUrl(url)) {
                String formattedUrl = formatUrl(url);
                TextViewCompat.setTextAppearance(viewHolder.itemSource, R.style.source_link);
                SpannableString content = new SpannableString(formattedUrl);
                content.setSpan(new UnderlineSpan(), 0, formattedUrl.length(), 0);
                viewHolder.itemSource.setText(content);
            } else {
                TextViewCompat.setTextAppearance(viewHolder.itemSource, R.style.source_text);
                viewHolder.itemSource.setText(url);
            }
        }
    }

    private String formatUrl(String urlString) {

        if (urlString == null) {
            return mParentActivity.getString(R.string.no_source);
        }

        try {
            URL url = new URL(urlString);
            return url.getHost();
        } catch (MalformedURLException e) {
            // urlString is not a proper url, just return it as is
            return urlString;
        }
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/adapter/EditRecipeAdapter.kt
package com.yeraygarcia.recipes.adapter

import android.graphics.Typeface.BOLD
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.RecyclerView
import android.text.Editable
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.TextWatcher
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import com.yeraygarcia.recipes.EditIngredientDialog
import com.yeraygarcia.recipes.EditIngredientDialog.Companion.TAG_DIALOG_EDIT_INGREDIENT
import com.yeraygarcia.recipes.EditStepDialog
import com.yeraygarcia.recipes.R
import com.yeraygarcia.recipes.database.entity.Recipe
import com.yeraygarcia.recipes.database.entity.RecipeStep
import com.yeraygarcia.recipes.database.entity.custom.UiRecipeIngredient
import com.yeraygarcia.recipes.viewmodel.RecipeDetailViewModel
import kotlinx.android.synthetic.main.item_edit_duration.view.*
import kotlinx.android.synthetic.main.item_edit_ingredient.view.*
import kotlinx.android.synthetic.main.item_edit_source.view.*
import kotlinx.android.synthetic.main.item_edit_step.view.*
import kotlinx.android.synthetic.main.item_header_generic.view.*
import kotlinx.android.synthetic.main.view_servings.view.*
import timber.log.Timber
import java.util.*


class EditRecipeAdapter(val activity: FragmentActivity, val viewModel: RecipeDetailViewModel) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val inflater = LayoutInflater.from(activity)

    var recipe: Recipe? = null
        set(value) {
            Timber.d("setRecipe(recipe)")
            field = value
            calculateItemCount()
            notifyDataSetChanged()
        }

    var ingredients: List<UiRecipeIngredient> = emptyList()
        set(value) {
            Timber.d("setIngredients(ingredients(${value.size}))")
            field = value
            calculateItemCount()
            notifyDataSetChanged()
        }

    var steps: List<RecipeStep> = emptyList()
        set(value) {
            Timber.d("setSteps(steps(${value.size}))")
            field = value
            calculateItemCount()
            notifyDataSetChanged()
        }

    private var itemCount: Int = -1

    fun EditText.onChange(cb: (String) -> Unit) {
        this.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                cb(s.toString())
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        })
    }

    internal inner class DurationViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val itemDuration: EditText = itemView.edittext_duration.apply {
            onChange { viewModel.saveDurationToDraft(it, recipe) }
        }
    }

    internal inner class SourceViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val itemSource: EditText = itemView.edittext_source.apply {
            onChange { viewModel.saveSourceToDraft(it, recipe) }
        }
    }

    internal inner class IngredientsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val itemIngredientName: TextView = itemView.textview_ingredient_name.apply {
            setOnClickListener {
                val ingredient = ingredients[getIngredientPosition(adapterPosition)]
                showEditIngredientDialog(ingredient.id)
            }
        }
    }

    internal inner class IngredientsHeaderViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        val itemServings: TextView = itemView.textview_servings
        val itemDecrease: ImageView = itemView.imageview_decrease_servings.apply {
            setOnClickListener {
                recipe?.let { recipe ->
                    recipe.decreasePortions()
                    viewModel.update(recipe)
                }
            }
        }
        val itemIncrease: ImageView = itemView.imageview_increase_servings.apply {
            setOnClickListener {
                recipe?.let { recipe ->
                    recipe.increasePortions()
                    viewModel.update(recipe)
                }
            }
        }
    }

    internal inner class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val itemTitle: TextView = itemView.textview_title
    }

    internal inner class StepsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val itemStepNumber: TextView = itemView.textview_step_number
        val itemStepDescription: TextView = itemView.textview_step_description.apply {
            setOnClickListener { showEditStepDialog(getStepAt(adapterPosition).id) }
        }
    }

    override fun getItemViewType(position: Int): Int {
        // there is always at least one ingredient item to show the "no ingredients" when empty
        val ingredientsCount = maxOf(ingredients.size, 1)

        return when {
            position == 0 -> VIEW_TYPE_DURATION
            position == 1 -> VIEW_TYPE_SOURCE
            position == 2 -> VIEW_TYPE_INGREDIENTS_HEADER
            position < ingredientsCount + 3 -> VIEW_TYPE_INGREDIENT
            position == ingredientsCount + 3 -> VIEW_TYPE_STEPS_HEADER
            else -> VIEW_TYPE_STEP
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_INGREDIENTS_HEADER -> IngredientsHeaderViewHolder(
                inflater.inflate(R.layout.item_header_ingredients, parent, false)
            )

            VIEW_TYPE_STEPS_HEADER -> HeaderViewHolder(
                inflater.inflate(R.layout.item_header_generic, parent, false)
            )


            VIEW_TYPE_DURATION -> DurationViewHolder(
                inflater.inflate(R.layout.item_edit_duration, parent, false)
            )

            VIEW_TYPE_SOURCE -> SourceViewHolder(
                inflater.inflate(R.layout.item_edit_source, parent, false)
            )

            VIEW_TYPE_INGREDIENT -> IngredientsViewHolder(
                inflater.inflate(R.layout.item_edit_ingredient, parent, false)
            )

            else -> StepsViewHolder(
                inflater.inflate(R.layout.item_edit_step, parent, false)
            )
        }
    }

    override fun getItemCount(): Int {
        if (itemCount == -1) {
            calculateItemCount()
        }
        return itemCount
    }

    private fun calculateItemCount() {
        itemCount = (if (recipe != null) 4 else 0) +
                maxOf(ingredients.size, 1) +
                maxOf(steps.size, 1)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        when (holder.itemViewType) {

            VIEW_TYPE_INGREDIENTS_HEADER -> (holder as IngredientsHeaderViewHolder).apply {
                recipe?.apply {
                    itemServings.text = String.format(Locale.getDefault(), "%d", portions)
                    itemDecrease.visibility = if (portions > 1) View.VISIBLE else View.GONE
                }
            }

            VIEW_TYPE_STEPS_HEADER -> (holder as HeaderViewHolder).itemTitle.setText(R.string.header_steps)

            VIEW_TYPE_DURATION -> recipe?.let { it ->
                val recipe = viewModel.recipeDraft ?: it
                val duration = formatDuration(recipe.durationInMinutes)
                (holder as DurationViewHolder).itemDuration.setText(duration)
            }

            VIEW_TYPE_SOURCE -> (holder as SourceViewHolder).itemSource.setText(recipe?.url)

            VIEW_TYPE_INGREDIENT -> (holder as IngredientsViewHolder).apply {
                if (ingredients.isNotEmpty()) {
                    getIngredientAt(position).apply {
                        itemIngredientName.text = makeBold(formattedQuantityAndUnit, toString())
                    }
                } else {
                    Timber.d("No ingredients")
                    // Covers the case of data not being ready yet.
                    itemIngredientName.setText(R.string.no_ingredients)
                }
            }

            else -> (holder as StepsViewHolder).apply {
                if (steps.isNotEmpty()) {
                    getStepAt(position).apply {
                        itemStepNumber.text = formatSortOrder(sortOrder)
                        itemStepDescription.text = description
                    }
                } else {
                    // Covers the case of data not being ready yet.
                    itemStepDescription.setText(R.string.no_steps)
                }
            }
        }
    }

    private fun makeBold(boldText: String, text: String): SpannableStringBuilder {
        val start = text.indexOf(boldText)
        val end = start + boldText.length
        val str = SpannableStringBuilder(text)
        str.setSpan(StyleSpan(BOLD), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        return str
    }

    private fun getIngredientPosition(position: Int): Int {
        return maxOf(position - 3, 0)
    }

    private fun getStepPosition(position: Int): Int {
        return maxOf(position - maxOf(ingredients.size, 1) - 4, 0)
    }

    private fun getIngredientAt(position: Int): UiRecipeIngredient {
        return ingredients[getIngredientPosition(position)]
    }

    private fun getStepAt(position: Int): RecipeStep {
        return steps[getStepPosition(position)]
    }

    private fun showEditIngredientDialog(ingredientId: UUID) {
        EditIngredientDialog.newInstance(ingredientId)
            .show(activity.supportFragmentManager, TAG_DIALOG_EDIT_INGREDIENT)
    }

    private fun showEditStepDialog(stepId: UUID) {
        EditStepDialog.newInstance(stepId)
            .show(activity.supportFragmentManager, TAG_DIALOG_EDIT_INGREDIENT)
    }

    private fun formatSortOrder(sortOrder: Int): String {
        return String.format(Locale.getDefault(), "%d", sortOrder)
    }

    private fun formatDuration(duration: Long): String {
        return String.format(Locale.getDefault(), "%d", duration)
    }
    
    companion object {
        private const val VIEW_TYPE_DURATION = 1
        private const val VIEW_TYPE_SOURCE = 2
        private const val VIEW_TYPE_INGREDIENTS_HEADER = 3
        private const val VIEW_TYPE_STEPS_HEADER = 4
        private const val VIEW_TYPE_INGREDIENT = 5
        private const val VIEW_TYPE_STEP = 6
    }
}